### PR TITLE
RF/FIX: Iterate over echo indices, not filenames, simplifying iteration logic

### LINKED ISSUE #2650: File name too long in slice time correction step on multi-echo rest runs
### What happened?

I tried running 9 subjects worth of data through fmriprep 21.0.0rc2 using a singluarity container built off of the docker image and all of them hung for 48 hours hitting the jobs wall time and dying. It looks like all of the jobs died in the slice timing correction step with `OSError: [Errno 36] File name too long:`. I think the processes hung because in addition to not being able to write the slice timing output, they were unable to write the error files for the same reason (File name too long).

### What command did you use?

```shell
singularity run --cleanenv /data/MBDU/singularity_images/fmriprep_v21.0.0rc2.simg /data/MBDU/mid_share/data/bids_defaced     $TMPDIR/out participant     --participant_label 20900     -w $TMPDIR/wrk     --nprocs $SLURM_CPUS_PER_TASK     --mem_mb $SLURM_MEM_PER_NODE     --fs-license-file /data/MBDU/singularity_images/license.txt     --output-spaces MNI152NLin2009cAsym:res-2 fsaverage     --dummy-scans 4 --bids-filter-file /data/MBDU/mid_share/data/derivatives/fmriprep/fmriprepv21.0.0rc2_test/bids_filter --use-aroma --error-on-aroma-warnings -vvv
```
```


### What version of fMRIPrep are you running?

fmriprep 21.0.0rc2

### How are you running fMRIPrep?

Singularity

### Is your data BIDS valid?

Yes

### Are you reusing any previously computed results?

No

### Please copy and paste any relevant log output.

shell
exception calling callback for <Future at 0x2aaab4b718b0 state=finished raised OSError>
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/nipype/pipeline/plugins/multiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/opt/conda/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 507, in run
    os.makedirs(outdir, exist_ok=True)
  File "/opt/conda/lib/python3.8/os.py", line 213, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/opt/conda/lib/python3.8/os.py", line 223, in makedirs
    mkdir(name, mode)
OSError: [Errno 36] File name too long: '/lscratch/27850734/wrk/fmriprep_wf/single_subject_20900_wf/func_preproc_ses_v1_task_rest_run_1_echo_1_wf/bold_stc_wf/_bold_file_..data..MBDU..mid_share..data..bids_defaced..sub-20900..ses-v1..func..sub-20900_ses-v1_task-rest_run-1_echo-2_bold.nii.gz_name_source_..data..MBDU..mid_share..data..bids_defaced..sub-20900..ses-v1..func..sub-20900_ses-v1_task-rest_run-1_echo-2_bold.nii.gz'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/concurrent/futures/process.py", line 239, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/opt/conda/lib/python3.8/site-packages/nipype/pipeline/plugins/multiproc.py", line 70, in run_node
    result["result"] = node.result
  File "/opt/conda/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 216, in result
    return _load_resultfile(
  File "/opt/conda/lib/python3.8/site-packages/nipype/pipeline/engine/utils.py", line 290, in load_resultfile
    if not results_file.exists():
  File "/opt/conda/lib/python3.8/pathlib.py", line 1388, in exists
    self.stat()
  File "/opt/conda/lib/python3.8/pathlib.py", line 1194, in stat
    return self._accessor.stat(self)
OSError: [Errno 36] File name too long: '/lscratch/27850734/wrk/fmriprep_wf/single_subject_20900_wf/func_preproc_ses_v1_task_rest_run_1_echo_1_wf/bold_stc_wf/_bold_file_..data..MBDU..mid_share..data..bids_defaced..sub-20900..ses-v1..func..sub-20900_ses-v1_task-rest_run-1_echo-2_bold.nii.gz_name_source_..data..MBDU..mid_share..data..bids_defaced..sub-20900..ses-v1..func..sub-20900_ses-v1_task-rest_run-1_echo-2_bold.nii.gz/slice_timing_correction/result_slice_timing_correction.pklz'
"""
```

### Additional information / screenshots

_No response_

### PR DESCRIPTION
**Pull request by effigies**
This PR:

1) Changes the source of iteration from `boldbuffer`/`meepi_echos` to `echo_index`, an `IdentityInterface` that is just `0` for single-echo workflows.
2) The echo index is passed to `bold_source` which indexes the list of BOLD images, and `validated_bold`, which indexes the list of BOLD images that have been validated in `initial_boldref_wf`.

Consequences:

1) Validated echos are now passed to STC/boldbuffer (was true for SE, but not for ME)
1) STC no longer has any dependency on number of echos
1) JoinNodes `joinsource` no longer depends on STC (non-)use
1) Iterable directories are now indexed by an integer instead of full file paths
1) `bold_bold_trans_wf` `name_source` no longer depends on number of echos

Fixes #2650.

### CODE REVIEW

File: fmriprep/workflows/bold/base.py [hunk 1]
```diff
@@ -488,43 +494,14 @@ def init_func_preproc_wf(bold_file, has_fieldmap=False):
...
-    elif not multiecho:  # SE and skip-STC
-        # fmt:off
-        workflow.connect([
-            (initial_boldref_wf, boldbuffer, [("outputnode.bold_file", "bold_file")]),
-        ])
-        # fmt:on
-    else:  # ME and skip-STC
-        # for meepi, iterate over all meepi echos to boldbuffer
-        boldbuffer.iterables = [
-            ("bold_file", bold_file),
```
Comment (effigies) [inline fmriprep/workflows/bold/base.py:525]: It looks like we had a ME/SE divergence here: For single-echo, we fixed up the header fields (matching qform/sform) but not for multi-echo. Will propose a fixup...

Comment (effigies) [reply fmriprep/workflows/bold/base.py:525]: The `validated_bold` `Select()` interface should now ensure that validated files are passed down for either ME or SE.

File: fmriprep/workflows/bold/base.py [hunk 2]
```diff
@@ -564,18 +545,18 @@ def init_func_preproc_wf(bold_file, has_fieldmap=False):
     # MAIN WORKFLOW STRUCTURE #######################################################
     # fmt:off
     workflow.connect([
+        # Prepare masked T1w image
         (inputnode, t1w_brain, [("t1w_preproc", "in_file"),
                                 ("t1w_mask", "in_mask")]),
+        # Select validated bold files per-echo
+        (initial_boldref_wf, validated_bold, [("val_bold.out_file", "inlist")]),
```
Comment (effigies) [inline fmriprep/workflows/bold/base.py:552]: This violates our usual rules, so I can create a new `outputnode` entry in niworkflows, assuming this PR is approved in principle.

**Comment by effigies** [comment]
@mgxd @oesteban This is ready for review. Updated the first post with a summary.

File: fmriprep/workflows/bold/base.py (hunk 2 above)
Comment (mgxd) [reply fmriprep/workflows/bold/base.py:552]: +1 to adding a new outputnode field.

**Review by mgxd — APPROVED** [review]
lgtm

**Pull request merged**